docs/dev_docs.md

```diff
@@ -1,3 +1,449 @@
+        # still be an entry for each agent
+        infos = {agent: {} for agent in self.agents}
+
+        return observations, rewards, dones, infos
+
+```
+
+## Wrappers
+
+A wrapper is a environment transformation which takes in an environment as input, and outputs a new environment which is similar to the inputted environment but some transformation or validation applied.
+
+### Conversion wrappers
+
+As we provide both the AEC API and the Parallel API, we also provide wrappers to convert back and forth between the two APIs.
+
+#### AEC to Parallel
+
+An environment can be converted from an AEC environment to a parallel environment with the `to_parallel` wrapper shown below. Note that this wrapper makes the following assumptions about the underlying environment:
+
+1. The environment steps in a cycle, i.e. it steps through every live agent in order.
+2. The environment does not update the observations of the agents except at the end of a cycle.
+
+Note that it also changes when rewards are allocated to agents. All rewards are given at the end of each cycle, which is slightly different than in the AEC model.
+
+```
+from pettingzoo.utils import to_parallel
+from pettingzoo.butterfly import pistonball_v3
+env = pistonball_v3.env()
+env = to_parallel(env)
+```
+
+#### Parallel to AEC
+
+Any parallel environment can be efficiently converted to an AEC environment with the `from_parallel` wrapper.
+
+```
+from pettingzoo.utils import from_parallel
+from pettingzoo.butterfly import pistonball_v3
+env = pistonball_v3.parallel_env()
+env = from_parallel(env)
+```
+
+### Utility wrappers
+
+We wanted our pettingzoo environments to be both easy to use and easy to implement. To combine these, we have a set of simple wrappers which provide input validation and other convenient reusable logic.
+
+* `BaseWrapper`: all AECEnv wrappers should inherit from this base class
+* `TerminateIllegalWrapper`: which handles illegal move logic for classic games
+* `CaptureStdoutWrapper`: which takes an environment which prints to terminal, and gives it an `ansi` render mode where it captures the terminal output and returns it as a string instead.
+* `AssertOutOfBoundsWrapper`: Asserts if the action given to step is outside of the action space. Applied in PettingZoo environments with discrete action spaces.
+* `ClipOutOfBoundsWrapper`: Clips the input action to fit in the continuous action space (emitting a warning if it does so). Applied to continuous environments in pettingzoo.
+* `OrderEnforcingWrapper`: Gives a sensible error message if function calls or attribute access are in a disallowed order, for example if step() is called before reset(), or .dones attribute is accessed before reset(), or if seed() and then step() is called before reset() is called again (reset must be called after seed()). Applied to all PettingZoo environments.
+
+All of the above can be applied to your environment with:
```

Comment:
```suggestion
Any of the above wrappers can be applied to your environment in the same way as the example below:
```
Is this trying to say that these specific 5 wrappers will be applied with the below code, or that you can apply anyone of these by following the example below?

Reply:
fixed